File: src/main/java/so/aporia/utils/user/render/animation/MessageAnim.java
package so.aporia.utils.user.render.animation;

/**
 * Per-message physics animation: alpha fade + horizontal slide-in.
 * Both driven by real spring oscillators — call {@link #tick()} each frame.
 */
public final class MessageAnim {

    private static final float ALPHA_K = 160f, ALPHA_D = 24f;
    private static final float SLIDE_K = 340f, SLIDE_D = 22f;

    /** Pixels to slide in from the left on first appear. */
    public static final float SLIDE_INITIAL = 12f;

    public final SpringSimulator alpha;
    public final SpringSimulator slideX;

    private long lastMs = System.currentTimeMillis();

    public MessageAnim(float initialAlpha) {
        alpha  = new SpringSimulator(ALPHA_K, ALPHA_D, initialAlpha);
        slideX = new SpringSimulator(SLIDE_K, SLIDE_D, initialAlpha > 0.5f ? 0f : SLIDE_INITIAL);
        alpha.setTarget(initialAlpha);
        slideX.setTarget(0f);
    }

    /** Advance springs by real elapsed time. Call once per render frame. */
    public void tick() {
        long now = System.currentTimeMillis();
        float dt = Math.min((now - lastMs) / 1000f, 0.05f);
        lastMs = now;
        alpha.update(dt);
        slideX.update(dt);
    }

    public void setAlphaTarget(float t) { alpha.setTarget(t); }

    public float alpha()  { return Math.max(0f, Math.min(1f, alpha.value())); }
    public float slideX() { return slideX.value(); }

    /** True once faded out and settled — safe to evict from cache. */
    public boolean isDead() {
        return alpha.getTarget() < 0.01f && alpha.isSettled();
    }
}

File: src/so/aporia/utils/user/render/animation/MessageAnim.kt
/*
 * Copyright (c) 2025-2026 Aporia.cc Project
 * Distributed under the Aporia.cc Software License Agreement v1.0
 * See LICENSE and COPYRIGHT files in the project root for full text.
 */

package so.aporia.utils.user.render.animation
import kotlin.math.max
import kotlin.math.min
import com.chaos.annotation.ChaosNative
/**
 * Per-message physics animation: alpha fade + horizontal slide-in.
 * Both driven by real spring oscillators — call [tick] each frame.
 */
@ChaosNative
class MessageAnim(initialAlpha: Float) {

    val alpha: SpringSimulator
    val slideX: SpringSimulator
    val slideY: SpringSimulator

    private var lastMs = System.currentTimeMillis()

    init {
        alpha = SpringSimulator(ALPHA_K, ALPHA_D, initialAlpha)
        slideX = SpringSimulator(SLIDE_K, SLIDE_D, if (initialAlpha > 0.5f) 0f else SLIDE_INITIAL)
        slideY = SpringSimulator(FLOAT_K, FLOAT_D, if (initialAlpha > 0.5f) 0f else FLOAT_INITIAL)
        alpha.setTarget(initialAlpha)
        slideX.setTarget(0f)
        slideY.setTarget(0f)
    }

    fun tick() {
        val now = System.currentTimeMillis()
        val dt = min((now - lastMs) / 1000f, 0.05f)
        lastMs = now
        alpha.update(dt)
        slideX.update(dt)
        slideY.update(dt)
    }

    fun setAlphaTarget(t: Float) { alpha.setTarget(t) }

    fun alpha(): Float = max(0f, min(1f, alpha.value()))
    fun slideX(): Float = slideX.value()
    fun slideY(): Float = slideY.value()

    fun isDead(): Boolean = alpha.getTarget() < 0.01f && alpha.isSettled()

    companion object {
        private const val ALPHA_K = 160f
        private const val ALPHA_D = 24f
        private const val SLIDE_K = 340f
        private const val SLIDE_D = 22f

        @JvmField
        val SLIDE_INITIAL = 12f

        private const val FLOAT_K = 120f
        private const val FLOAT_D = 20f

        @JvmField
        val FLOAT_INITIAL = 8f
    }
}